### PR TITLE
Adds the NPCPool subsystem. (For Bots and SNPCs)

### PR DESCRIPTION
**Pull request by sawu-tg**
- NPCPool will co-ordinate and organize all SNPCs, making them stronger and more intelligent.
- SNPCs can use the botPool system to co-ordinate and delegate with each other.
- SNPCs have been given factions to allow interoperability between certain types of SNPCs.
- Bots (machinery/bot) can use the system to for basic help and coordination.

### CODE REVIEW

**Comment by Jalleo** [comment]
:+1: Does this mean you can call for multiple cleaning bots in the same call?

**Comment by sawu-tg** [comment]
@Jalleo It works mainly for SNPCs, but any NPC that uses the SNPC system will work with it.

It uses SNPC specific functions to check the pooled NPCs and whether they need assistance.

**Comment by Jalleo** [comment]
Would you be willing to work with gun to expand this to bots?

**Comment by sawu-tg** [comment]
@Jalleo I can attempt to make it work for bots, but it would require a whole set of different variables.

**Comment by Jalleo** [comment]
Gotcha I would love when a expansion of the bot control comes up for a PR at some point down the road (what epse are you currently planning to do with your npcs?

**Comment by sawu-tg** [comment]
@Jalleo I want to expand NPCs to be able to do simple jobs, as well as other tasks. Maybe even turn them into a menial labor force for the station.

**Comment by Jalleo** [comment]
Nice we may be able to create npcs able to deal with the station in a emergency bu being the new ert or deathsquad (This would change everything)

**Comment by sawu-tg** [comment]
@Jalleo I've added some basic support for regular bots now.

**Comment by Iamgoofball** [comment]
Hivemind SNPCS :+1:

**Comment by Dorsisdwarf** [comment]
Make it so that if there are less than 15 people when the round starts it spawns a bunch of these to keep the station feeling lively!

**Comment by optimumtact** [comment]
Docs and rocks :+1:

**Comment by Jalleo** [comment]
@Dorsisdwarf only once they are able to interface and understand at least a few things

File: code/controllers/subsystem/botpool.dm [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+	// 1. Bot proximity to task target: if too far, delegate, if close, coordinate
+	// 2. Bot Health/status: check health with bots in local area, if their health is higher, delegate task to them, else coordinate
+	// 3. Process delegation: if a bot (or bots) has been delegated, assign them to the task.
+	// 4. Process coordination: if a bot(or bots) has been asked to coordinate, assign them to help.
+	// 5. Do all assignments: goes through the delegated/coordianted bots and assigns the right variables/tasks to them.
+	var/npcCount
+
+	//bot handling
+	for(npcCount = 1; npcCount < botPool_l_non.len; ++npcCount)
+		var/obj/machinery/bot/check = botPool_l_non[npcCount]
```
Comment (Carn) [inline code/controllers/subsystem/botpool.dm]: This is the slowest way you can write a for loop. There's a reason all the others do not do it this way.
Keep it consistent.

**Comment by Carn** [comment]
Also why can't this be merged into the bot subsystem?

**Comment by optimumtact** [comment]
It's kind of misnamed since it's not for bots, it's for NPC characters.

**Comment by Carn** [comment]
Ok, fair enough. Could we rename it if possible.

**Comment by sawu-tg** [comment]
@Dorsisdwarf @Jalleo Pretty much what Jalleo said, they can "use" objects, and even build things, but its a monkeys and typewriters situation, there's no specific modules as of yet to allow them to fully do a job.

They can act as admin events and general "filler" but unfortunately, right now, their jobs consist of hitting things with things.

File: code/controllers/subsystem/botpool.dm (hunk 1 above)
Comment (sawu-tg) [inline code/controllers/subsystem/botpool.dm]: Sorry, people keep confusing me with what is and is not the fastest way. I'll fix that up post-haste.

**Comment by sawu-tg** [comment]
@optimumtact [Click here for documentation](https://dl.dropboxusercontent.com/u/1513925/ss13/npcpool_docs.txt)

**Comment by Miauw62** [comment]
i can confirm that this is hilarious.

**Comment by Jalleo** [comment]
@sawu-tg with the botpool and bots are they checking if there is a way to access the area by the bot before the request is placed into the pool?

Secondly would it be possible for people to add or remove new tasks in the botpool for the bots being added at a later date for ANOTHER PR? (I would like this PR to be merged before more feature creep is added)

**Comment by Dorsisdwarf** [comment]
In response to "Their jobs consist of hitting things with things": Do I sense a replacement for security? :grinning:

**Comment by sawu-tg** [comment]
@Jalleo All procs are done via commands on the bot, not the pool. Bots (obj/machinery) use all access and thus can access anywhere. SNPCs will delegate the task across each other until it reaches a valid bot or it is abandoned.

I'm not sure what you mean about "tasks". You could add more clauses in which the pool triggers, but basically all it does right now is co-ordinate things, everything else is done in the bots and SNPCs themselves.

**Comment by Jalleo** [comment]
gotcha I think their access got changed to not access everything now though ill check with gun_hog

**Comment by dannno** [comment]
this is the future

**Comment by sawu-tg** [comment]
@Jalleo It uses the AI's call_bot command.

**Comment by Jalleo** [comment]
Nice

**Comment by Iamgoofball** [comment]
???

**Pull request closed without merging**